### DISCUSSION TITLE
How to import training package as new project with reference to real videos and not the training package itself?

### DISCUSSION THREAD

**Discussion by rfkova** [talmolab/sleap#1519] [Help!]
Hello!

This is simple but I can't figure it out right now... help!

I have a project that someone else was working on and I don't have access to the original project file. I do have access to the most recent training package they exported and the original videos. I want to get the labels out of the training package and back into a project file with real videos assigned, not the internal reference to the training package which doesn't contain the original videos.

When I try to Merge the training package into an empty project with the original videos, it just creates new videos that reference the training package (.slp not .mp4). If I try to reassign the video from the slp to mp4, it only let's me look for .slp file extensions, not the .mp4 videos I need to reassign them to.

The annoying part is I've figured out how to do this before and recall it was not complicated, just slightly unintuitive and now I'm very lost!

Thanks!
Ryan

**Comment by roomrys** [comment]
Hi @rfkova,

Is there any way you can get access to the slp? This is a bit outside of the current functionality of the `.pkg.slp` which was intended for exporting to train, but never importing or editing. I know you said you've done this once before, but I am not quite sure how you did it....

Thanks,
Liezl

**Reply by rfkova** [reply]
Hi @roomrys,

Hmmm, fascinating. 

I probably can lay hands on the original but in the midst of a crunch and having some evidence I managed before I went ahead and audited the labels all of yesterday assuming I could be clued in to how to convert them later. Bummer!

I probably can get the original file in a few days, but I don't know that that helps me import the audit I did on the training labels. 

I guess give it another quick think and if there is no solution that is apparent we can just close the issue in defeat. I don't think it make sense to script a solution for something like this, which is clearly unintended functionality. It's so irritating to have the evidence of a prior conversion in front of me, haha. 

Thanks,
Ryan

**Reply by rfkova** [reply]
Like, I can figure out which video is which. There ane only 20 of them. If that helps at all. The training package lists the source as .slp thought which is what I'd like to change.

**Reply by roomrys** [reply]
Ok, it will have to be done through a script, let me write something for you.

**Comment by roomrys** [comment]
Hi @rfkova, 

Here is a small script for you to make edits to in the `if __name__ == "__main__"` portion. 

I've ran through it and verified that the instances are on the correct frames and that the rest of the video frames load. It is a [simple script](https://gist.github.com/roomrys/d7efe93fe98fb786088a2ca6933edf75), so you will need to manually order your videos in the correct order and have the exact same number of videos listed for replacement as are in the package slp. There will be some debugging messages in the terminal to help sort this out. Let me know how it works for you!

Thanks,
Liezl

```python
import logging

import sleap
from sleap.io.video import MediaVideo

logger = logging.getLogger(__name__)


def main(path_pkg_slp, paths_to_videos, output_slp_path="slp_from_package.slp"):
    """Replace HDF5 backend with MediaVideo backend for videos in package.

    Args:
        path_pkg_slp: Path to SLEAP package file.
        paths_to_videos: List of paths to videos, must be the exact same number of
            videos as in the package. Assumes MediaVideo backend.
        output_slp_path: Path to output SLEAP package file.

    Returns:
        Path to output SLEAP package file.
    """

    labels_pkg_slp = sleap.load_file(path_pkg_slp)
    logger.info(f"Num videos in package: {len(labels_pkg_slp.videos)}")
    logger.info(f"Num videos to replace: {len(paths_to_videos)}")

    if len(labels_pkg_slp.videos) != len(paths_to_videos):
        raise ValueError(
            "Number of videos in package does not match number of videos in list. "
            "Perhaps only the user labeled frames (and related videos) were exported?"
        )

    for video, path_to_video in zip(labels_pkg_slp.videos, paths_to_videos):
        new_video_backend = MediaVideo(filename=path_to_video, grayscale=True)
        video.backend = new_video_backend

    labels_pkg_slp.save(filename=output_slp_path)
    logger.info(f"Saved SLEAP package to {output_slp_path}")

    return output_slp_path


if __name__ == "__main__":

    from pathlib import Path

    from sleap.gui.app import main as sleap_label

    base_path = r"D:\social-leap-estimates-animal-poses\datasets\drosophila-melanogaster-courtship\drosophila-melanogaster-courtship"
    path_to_pkg_slp = Path(base_path, "courtship_labels copy.pkg.slp").as_posix()
    paths_to_videos = [
        Path(base_path, "dmc_pad-w.mp4").as_posix(),
        Path(base_path, "dmc_upsample.mp4").as_posix(),
    ]
    output_slp_path = r"slp_from_package.slp"

    path_to_new_slp = main(path_to_pkg_slp, paths_to_videos, output_slp_path)
    sleap_label([path_to_new_slp])

```

**Reply by rfkova** [reply]
@roomrys,

Thank you so much! More than I could have asked for. 

Sadly, I have to run for the weekend, but I'll let you know early next week how this panned out :D

**Comment by rfkova** [comment]
@roomrys, 

Hi... so... I haven't tried your script because I think I stumbled into the answer of how I got it working again earlier this week but needed access to the original videos which I got today... I think it works. Let me give you the sequence of events so you can follow this up if you like...

Steps:
1. Get someone else to do some labeling for you (be charming, bring flowers, idk, do what you have to).
2. Lose access to their original .slp file.
3. Retrieve a training package exported from a video with labels.
4. Realize you can't replace the listed .slp video sources -> panic -> ask for help -> soldier on while awaiting feedback.
5. Open the training package as a project in it's own right.
6. Save it to a new file to preserve the original.
7. Edit the labels therein to suit new criteria (moving them around and occasionally adding and deleting instances as convenience dictates, no changing skeleton), saving as you go.
8. Close the file and terminal for the day and go to sleap... I mean, sleep.
9. Reopen the saved project generated from the training package... OMG IT IS PROMPTING ME FOR VIDEO FILES WHICH IT DIDN'T DO WHEN I OPENED THE TRAINING PACKAGE ITSELF.
10. Save file and hope everything is ok.
11. Feel bad that maybe you got someone very kind and important to script something they ultimately didn't need to but rather than apologize outright make a long list of steps with that sentiment embedded therein.
12. Profit?

I think I'm going to go about my life as though everything is fixed, but I'll post in this thread again if there is a problem and will try out the script to correct matters, should they need correcting. Looks good so far though.

THANK YOU SO MUCH!
Ryan
